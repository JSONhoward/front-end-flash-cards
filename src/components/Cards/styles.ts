import styled from 'styled-components';

export const StyledCards = styled('div')`
position: relative;
display: flex;
flex-direction: column;
align-items: center;
justify-content: center;
height: 100%;

h1 {
    color: whitesmoke;
}
`

export const QuestionCount = styled('p')`
color: whitesmoke;
margin: 1rem;
`

export const CardsContainer = styled('div')`
position: relative;
height: 40vw;
width: 50vw;
min-height: 250px;
min-width: 280px;
`

export const CardCategory = styled('div')`
position: absolute;
left: 0;
bottom: -1rem;
display: flex;
flex-direction: column;
align-items: center;
justify-content: flex-end;
height: 2rem;
/* width: 10rem; */
font-size: 1.5vw;
color: whitesmoke;
text-shadow: 2px 2px 10px rgba(0,0,0,.5);

select {
    display: block;
    height: 2rem;
    width: 6rem;
    text-align: center;
    margin: 0 5px;
    font-size: 1.5rem;
    padding-left: 5px;
    box-shadow: 2px 2px 10px rgba(0,0,0,.5);
    background-color: whitesmoke;
}

select:hover {
    border: 2px solid #888;
    cursor:pointer;
}

@media screen and (max-width: 600px) {
    font-size: .75rem;

    select {
        height: 1.5rem;
        font-size: 1rem;
        padding: 2px;
    }
}
`

export const NextCard = styled('div')`
position: absolute;
bottom: -.5rem;
right: 0;
display: flex;
align-items: center;
justify-content: flex-end;
font-size: 3vw;
color: whitesmoke;
text-shadow: 2px 2px 10px rgba(0,0,0,.5);
cursor: pointer;

@media screen and (min-width: 600px) {
    font-size: 1.5rem
}
`

export const ShowAnswer = styled('div')<{answer: boolean}>`
position: absolute;
bottom: -.5rem;
left: 50%;
transform: translateX(-50%);
display: flex;
align-items: center;
color: ${({answer}) => answer ? 'grey' : 'whitesmoke'};
font-size: 3vw;
cursor: ${({answer}) => answer ? 'not-allowed' : 'pointer'};
text-shadow: 2px 2px 10px rgba(0,0,0,.5);
text-decoration: ${({answer}) => answer ? 'line-through' : 'none'};
font-style: ${({answer}) => answer ? 'italic' : 'none'};
transition: color .25s ease-in-out;

&:hover {
    color: ${({answer}) => answer ? 'grey' : 'green'};
}

@media screen and (min-width: 600px) {
    font-size: 1.5rem
}
`