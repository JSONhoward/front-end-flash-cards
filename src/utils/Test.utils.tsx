import * as React from 'react'
import { RecoilRoot } from 'recoil'

export const RecoilWrapper: React.FC = ({children}) => {
    return (
        <RecoilRoot>
            {children}
        </RecoilRoot>
    )
}