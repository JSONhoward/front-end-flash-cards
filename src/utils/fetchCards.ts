export const fetchCards = async (): Promise<FirebaseDocArray> => {
    const response = await fetch(process.env.REACT_APP_GRAPHQL_URL as string, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'X-REQUEST-TYPE': 'GraphQL'
        },
        body: JSON.stringify({
            query: fetchAllCardsQuery
        })
    })
    const data = await response.json()
    const {data: {cards}} = data

    for (let i = cards.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [cards[i], cards[j]] = [cards[j], cards[i]]
    }

    return cards
}

const fetchAllCardsQuery = `
query {
    cards {
        id
        Category
        q
        code
        a
    }
}
`