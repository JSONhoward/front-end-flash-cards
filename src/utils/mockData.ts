export const mockQuestionArray: FirebaseDocArray = [
    {
        id: '1',
        Category: 'General',
        q: 'When building a new web site or maintaining one, can you explain some techniques you have used to increase performance?',
        a: 'some answer'
    },
    {
        id: '2',
        Category: 'General',
        code: 'some code',
        q: 'Test question 2',
        a: 'some answer'
    },
    {
        id: '3',
        Category: 'HTML',
        code: 'some code',
        q: 'Test question 3',
        a: 'some answer'
    },
    {
        id: '4',
        Category: 'Javascript',
        q: 'Test question 4',
        a: 'some answer'
    },
    {
        id: '5',
        Category: 'CSS',
        q: 'Test question 5',
        a: 'some answer'
    },
    {
        id: '6',
        Category: 'Networking',
        q: 'Test question 6',
        a: 'some answer'
    },
    {
        id: '7',
        Category: 'Testing',
        q: 'Test question 7',
        a: 'some answer'
    },
    {
        id: '8',
        Category: 'Coding',
        q: 'Test question 8',
        a: 'some answer'
    },
    {
        id: '9',
        Category: 'General',
        q: 'Test question 9',
        a: 'some answer'
    },
    {
        id: '10',
        Category: 'General',
        q: 'Test question 10',
        a: 'some answer'
    }
]