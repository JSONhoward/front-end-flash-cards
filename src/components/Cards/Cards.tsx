import React, { useEffect, useState } from 'react'
import { useRecoilState, useRecoilValue } from 'recoil'

import { CardOffset } from '../../store/atoms';
import { FiveQuestions, TotalQuestions } from '../../store/selectors';
import Card from './Card/Card';
import { CardCategory, CardsContainer, NextCard, ShowAnswer, StyledCards, QuestionCount } from './styles';

const Cards = () => {
    const [cardOffset, setCardOffset] = useRecoilState(CardOffset)
    const [category, setCategory] = useState('all')
    const totalQuestions = useRecoilValue(TotalQuestions)
    const cardsArray = useRecoilValue(FiveQuestions({ cardOffset, category }))
    const [removeCard, setRemoveCard] = useState(false)
    const [showAnswer, setShowAnswer] = useState(false)
    const [loading, setLoading] = useState(true)

    useEffect(() => {
        if (cardsArray[0]?.id !== '0') setLoading(false)
    }, [cardsArray])

    const handleRemoveCard = () => {
        setRemoveCard(true)
        setShowAnswer(false)
        setLoading(true)
        setTimeout(async () => {
            setCardOffset(prev => prev + 1)
            setRemoveCard(false)
            setLoading(false)
        }, 1000)
    }

    const handleDropdown = (e: React.SyntheticEvent) => {
        setCategory((e.target as HTMLSelectElement).value)
    }

    let cardStack = cardsArray?.map((el, i) => {
        return (
            <Card
                data-testid={`cardTest${i}`}
                key={el.id}
                loading={loading}
                question={el.q}
                code={el.code}
                answer={el.a}
                remove={i === 0 ? removeCard : false}
                show={showAnswer}
                offset={i}
            />
        )
    }).reverse()

    return (
        <StyledCards>
        <h1>Front End Developer Interview Questions</h1>
        <QuestionCount>Question# {cardOffset + 1} of {totalQuestions}</QuestionCount>
            <CardsContainer>
                {cardStack}
                <CardCategory>
                    <label htmlFor='category'>Category:</label>
                    <select name='category' onChange={handleDropdown}>
                        <option value='all'>All</option>
                        <option value='General'>General</option>
                        <option value='HTML'>HTML</option>
                        <option value='CSS'>CSS</option>
                        <option value='Javascript'>Javascript</option>
                        <option value='Networking'>Networking</option>
                        <option value='Testing'>Testing</option>
                    </select>
                </CardCategory>
                <ShowAnswer data-testid='showAnswer' answer={showAnswer} onClick={() => setShowAnswer(true)}><p>Show Answer</p></ShowAnswer>
                <NextCard data-testid='nextCard' onClick={handleRemoveCard}>
                    Next Card
                </NextCard>
            </CardsContainer>
        </StyledCards>
    )
}

export default Cards
