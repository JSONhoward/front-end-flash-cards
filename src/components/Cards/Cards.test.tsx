import React from 'react'
import { create } from 'react-test-renderer'
import { fireEvent, cleanup, render, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import { RecoilWrapper } from '../../utils/Test.utils'
import Cards from './Cards'

afterEach(cleanup)

describe('Cards Tests', () => {
    it('Matches Snapshot', () => {
        const component = create(<RecoilWrapper><Cards /></RecoilWrapper>)

        expect(component.toJSON()).toMatchSnapshot()
    })

    it('Shows answer style changes upon clicking "Show Answer"', () => {
        render(<RecoilWrapper><Cards /></RecoilWrapper>)

        const showAnswerButton = screen.getByTestId('showAnswer')

        expect(showAnswerButton).not.toHaveStyleRule('text-decoration', 'line-through')
        expect(showAnswerButton).toHaveStyleRule('color', 'whitesmoke')
        fireEvent.click(showAnswerButton)
        expect(showAnswerButton).toHaveStyleRule('text-decoration', 'line-through')

    })

    it('Increments Question Number', async () => {
        render(<RecoilWrapper><Cards /></RecoilWrapper>)

        console.log(screen.getByTestId('questionCount').textContent)
    })
})