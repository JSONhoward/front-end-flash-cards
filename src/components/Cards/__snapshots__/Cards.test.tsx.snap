// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards Tests Matches Snapshot 1`] = `
.c3 {
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  height: 30vw;
  width: 50vw;
  min-height: 168px;
  min-width: 280px;
  padding: 4rem;
  background-color: whitesmoke;
  box-shadow: 2px 2px 10px rgba(0,0,0,.5);
  font-family: 'Permanent Marker',cursive;
  font-size: 1.5rem;
}

.c3 p {
  text-align: center;
}

.c2 {
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  height: 30vw;
  width: 50vw;
  min-height: 168px;
  min-width: 280px;
  padding: .5rem;
  background-color: whitesmoke;
  box-shadow: 2px 2px 10px rgba(0,0,0,.5);
  -webkit-animation: none;
  animation: none;
  font-family: 'Permanent Marker',cursive;
  font-size: 1.5rem;
}

.c2 p {
  text-align: center;
  font-size: 3vw;
  white-space: pre-wrap;
}

.c2 code {
  background-color: black;
  max-height: 50%;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  min-width: 75%;
  color: white;
  font-family: monospace;
  font-size: .5rem;
  white-space: pre-wrap;
  overflow: auto;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 5rem;
  width: 50vw;
  min-width: 280px;
}

.c4 {
  color: whitesmoke;
}

.c1 {
  position: relative;
  height: 40vw;
  width: 50vw;
  min-height: 250px;
  min-width: 280px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  height: 2rem;
  font-size: 1.5vw;
  color: whitesmoke;
  text-shadow: 2px 2px 10px rgba(0,0,0,.5);
}

.c6 select {
  display: block;
  height: 2rem;
  width: 6rem;
  text-align: center;
  margin: 0 5px;
  font-size: 1.5rem;
  padding-left: 5px;
  box-shadow: 2px 2px 10px rgba(0,0,0,.5);
  background-color: whitesmoke;
}

.c6 select:hover {
  border: 2px solid #888;
  cursor: pointer;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  font-size: 3vw;
  color: whitesmoke;
  text-shadow: 2px 2px 10px rgba(0,0,0,.5);
  cursor: pointer;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: whitesmoke;
  font-size: 3vw;
  cursor: pointer;
  text-shadow: 2px 2px 10px rgba(0,0,0,.5);
  -webkit-text-decoration: none;
  text-decoration: none;
  font-style: none;
  -webkit-transition: color .25s ease-in-out;
  transition: color .25s ease-in-out;
}

.c7:hover {
  color: green;
}

@media screen and (min-width:600px) {
  .c2 p {
    font-size: 1.5rem;
  }

  .c2 code {
    font-size: 1.25rem;
  }
}

@media screen and (max-width:600px) {
  .c6 {
    font-size: .75rem;
  }

  .c6 select {
    height: 1.5rem;
    font-size: 1rem;
    padding: 2px;
  }
}

@media screen and (min-width:600px) {
  .c8 {
    font-size: 1.5rem;
  }
}

@media screen and (min-width:600px) {
  .c7 {
    font-size: 1.5rem;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
      offset={0}
    >
      <div
        className="c3"
      >
        Loading...
      </div>
    </div>
  </div>
  <p
    className="c4"
  >
    Question# 
    1
     of 
    1
  </p>
  <div
    className="c5"
  >
    <div
      className="c6"
    >
      <label
        htmlFor="category"
      >
        Category:
      </label>
      <select
        name="category"
        onChange={[Function]}
      >
        <option
          value="all"
        >
          All
        </option>
        <option
          value="General"
        >
          General
        </option>
        <option
          value="HTML"
        >
          HTML
        </option>
        <option
          value="CSS"
        >
          CSS
        </option>
        <option
          value="Javascript"
        >
          Javascript
        </option>
        <option
          value="Networking"
        >
          Networking
        </option>
        <option
          value="Testing"
        >
          Testing
        </option>
      </select>
    </div>
    <div
      className="c7"
      data-testid="showAnswer"
      onClick={[Function]}
    >
      <p>
        Show Answer
      </p>
    </div>
    <div
      className="c8"
      data-testid="nextCard"
      onClick={[Function]}
    >
      Next Card
    </div>
  </div>
</div>
`;
